coercion.ts

```diff
@@ -113,121 +110,68 @@ function coerceFile(file: unknown) {
+  rewrap = false,
 ) {
-  // @ts-expect-error
-  const { coerced, schema: wrapSchema } = enableTypeCoercion(type.wrapped);
+  const { transformAction, schema: wrapSchema } = enableTypeCoercion(
+    // @ts-expect-error
+    type.wrapped,
+  );
 
-  if (coerced) {
+  if (transformAction) {
     // `expects` is required to generate error messages for `TupleSchema`, so it is passed to `UnkonwSchema` for coercion.
     const unknown = { ...valibotUnknown(), expects: type.expects };
-    if (type.async) {
-      switch (schemaType) {
-        case "nullish":
-          return {
-            coerced: false,
-            schema: nullishAsync(pipeAsync(unknown, wrapSchema.pipe[1], type)),
-          };
-        case "optional":
-          return {
-            coerced: false,
-            schema: optionalAsync(pipeAsync(unknown, wrapSchema.pipe[1], type)),
-          };
-        default:
-          return {
-            coerced,
-            schema: pipeAsync(unknown, wrapSchema.pipe[1], type),
-          };
-      }
-    }
-    switch (schemaType) {
-      case "nullish":
-        return {
-          coerced: false,
-          schema: nullish(pipe(unknown, wrapSchema.pipe[1], type)),
-        };
-      case "optional":
-        return {
-          coerced: false,
-          schema: optional(pipe(unknown, wrapSchema.pipe[1], type)),
-        };
-      default:
-        return {
-          coerced,
-          schema: pipe(unknown, wrapSchema.pipe[1], type),
-        };
+    const schema = type.async
+      ? pipeAsync(unknown, transformAction, type)
+      : pipe(unknown, transformAction, type);
+
+    if (rewrap) {
+      return {
+        transformAction: undefined,
+        schema: type.reference(schema),
+      };
     }
+
+    return { transformAction, schema };
   }
 
   const wrappedSchema = {
-    ...originalSchema,
-    // @ts-expect-error
-    wrapped: enableTypeCoercion(originalSchema.wrapped).schema,
+    ...type,
+    wrapped: wrapSchema,
   };
 
   return {
-    coerced: false,
-    schema: generateReturnSchema(type, wrappedSchema),
+    transformAction: undefined,
+    schema: wrappedSchema,
   };
 }
 
 /**
  * Reconstruct the provided schema with additional preprocessing steps
  * This coerce empty values to undefined and transform strings to the correct type
  */
+export function enableTypeCoercion<
+  T extends GenericSchema | GenericSchemaAsync,
+>(
+  type: T,
+): {
+  transformAction: TransformAction<unknown, unknown> | undefined;
+  // If we use just `T` for type of `schema`, `enabltTypeCoercion<GenericSchema<string, string>>` will return
+  // `GenericSchema<string, string>`. However, we want it to return `GenericSchema<unknown, string>`.
+  schema: T extends GenericSchema<unknown, infer Output, infer Issue>
+    ? GenericSchema<unknown, Output, Issue>
+    : T extends GenericSchemaAsync<unknown, infer OutputAsync, infer IssueAsync>
+      ? GenericSchemaAsync<unknown, OutputAsync, IssueAsync>
+      : never;
+};
```

Comment:
I separated external (to be exported) signature and internal (to be referenced in function body) signature of `enableTypeCoercion` function to pass type-checking without using `@ts-expect-error` comments on `return` statements.